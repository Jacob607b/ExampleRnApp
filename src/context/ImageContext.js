import { createContext } from "react";

export const ImageContext = createContext({ Images: [null] });

export default ImageContext;
