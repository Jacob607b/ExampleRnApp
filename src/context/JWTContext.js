import { createContext } from "react";

export const JWTContext = createContext({ Jwt: undefined });

export default JWTContext;
