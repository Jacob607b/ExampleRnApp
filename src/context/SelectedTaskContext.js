import { createContext } from "react";

export const SelectedTaskContext = createContext({selectedTask : []});

export default SelectedTaskContext;