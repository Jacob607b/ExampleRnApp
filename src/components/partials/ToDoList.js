import React, { useContext, useCallback, useState } from "react";
import {
  View,
  Text,
  StyleSheet,
  ActivityIndicator,
  TouchableOpacity,
  Modal,
  ScrollView,
} from "react-native";
import { Button } from "react-native-paper";
import { getAuth } from "firebase/auth";
import { useQuery } from "@tanstack/react-query";
import ListItem from "./children/ListItem";
import Animated, { FadeInDown, FadeOut } from "react-native-reanimated";
import { IDContext } from "../../context/IDContext";
import { SelectedTaskContext } from "../../context/SelectedTaskContext";

export default function ToDoList({ showToggleButton }) {
  const { ID } = useContext(IDContext);
  const { setSelectedTask } = useContext(SelectedTaskContext);
  const [visible, setVisible] = useState(true);

  const containerStyle = {
    backgroundColor: "#f2f2f2",
    padding: 20,
    flexDirection: "column",
    justifyContent: "center",
    alignItems: "center", // Change alignItems to stretch
    textAlign: "center",
    height: "60%",
    width: "100%"
  };
  

  const fetchToDo = useCallback(
    async (ID) => {
      const auth = getAuth();
      const user = auth.currentUser;
      const token = await user.getIdToken();
      const requestOptions = {
        method: "POST",
        headers: {
          "Content-Type": "application/json",
          Authorization: `Bearer ${token}`,
        },
        body: JSON.stringify({ ID }),
      };

      const res = await fetch(
        "http://10.0.2.2:5001/vcs-reads-dev2/us-central1/app/toDo",
        requestOptions
      );
      const data = await res.json();
      return data;
    },
    [ID]
  );

  const { data, isLoading, isError } = useQuery({
    queryKey: ["todos", ID],
    queryFn: () => fetchToDo(ID),
    staleTime: Infinity,
  });

  const hideModal = useCallback(() => {
    setVisible(false);
    showToggleButton();
  }, [showToggleButton]);

  const handleSelectedTask = async (item) => {
    await setSelectedTask(item);
    hideModal();
  };

  if (isLoading) return <ActivityIndicator size="large" color={"#ed3b49"} testID="spinner"/>;
  if (isError) return <Text>Error fetching data</Text>;
  if (data.length === 0) return <Text>No Assigned Tasks</Text>;

  return (
    <Animated.View exiting={FadeOut} style={styles.backgroundColor} testID="toDoList">
      <Modal
        visible={visible}
        transparent={true}
        onRequestClose={hideModal}
        presentationStyle="overFullScreen"
      >
        <View style={styles.modalBackground}>
          <View style={containerStyle}>
            <View style={styles.modalHeader}>
              <Text style={styles.headerText}>Please Select A Task</Text>
            </View>

            <Animated.FlatList
              entering={FadeInDown.springify().mass(1)}
              style={{ flex: 1 }}
              data={data.filter((task) => !task.completed)}
              keyExtractor={(item) =>
                `${item.Unit}_${item.Type}_${item.Description}`
              }
              renderItem={({ item }) => (
                <ScrollView>
                  <ListItem
                    item={item}
                    handleSelectedTaks={() => handleSelectedTask(item)}
                  />
                </ScrollView>
              )}
            />
            <TouchableOpacity
              onPress={hideModal}
              testID="Dismiss"
              style={{ marginBottom: 5, marginTop: 10 }}
            >
              <Button mode="elevated" compact={true}>
                Close
              </Button>
            </TouchableOpacity>
          </View>
        </View>
      </Modal>
    </Animated.View>
  );
}

const styles = StyleSheet.create({
  container: {
    borderWidth: 1,
    borderColor: "black",
    width: "80%",
    marginRight: 8,
    padding: 3,
    elevation: 3,
    marginBottom: 5,
  },
  backgroundColor: {
    color: "#4D626C",
  },
  modalBackground: {
    flex: 1,
    backgroundColor: "rgba(0, 0, 0, 0.5)",
    justifyContent: "center",
    alignItems: "center",
  },
  modalHeader: {
    justifyContent: "center",
    alignItems: "center",
    marginBottom: 5,
    marginTop: 5,
    backgroundColor: "#008dc9",
    height: "10%",
    width: "100%",
  },
  headerText: {
    fontWeight: "bold",
    fontSize: 15,
    color: "white",

  },
});
