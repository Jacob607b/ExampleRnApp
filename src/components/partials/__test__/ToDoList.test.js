import React from "react";
import ToDoList from "../ToDoList";
import { render, act, screen, fireEvent } from "@testing-library/react-native";
import { IDContext } from "../../../context/IDContext";
import SelectedTaskContext from "../../../context/SelectedTaskContext";
import {
  QueryClient,
  QueryClientProvider,
  useQuery,
} from "@tanstack/react-query";
jest.mock("@tanstack/react-query", () => ({
  ...jest.requireActual("@tanstack/react-query"),
  useQuery: jest.fn().mockReturnValue(),
}));

describe("ToDoList Tests", () => {
  jest.mock("../../../context/SelectedTaskContext", () => ({
    ...jest.requireActual("../../../context/SelectedTaskContext"),
    useContext: jest.fn(() => ({
      setSelectedTask: jest.fn(),
    })),
  }));


  it("No Task Assigned", async () => {
    const queryClient = new QueryClient();
    const mockData = [];
    await act(() => {
      useQuery.mockReturnValueOnce({
        data: mockData,
        isLoading: false,
        isError: false,
      });
    });

    const setSelectedTask = jest.fn();

    const data = [];
    global.fetch = jest.fn().mockResolvedValueOnce({
      json: jest.fn().mockResolvedValueOnce(data),
    });
    render(
      <QueryClientProvider client={queryClient}>
        <IDContext.Provider value={{ ID: "1" }}>
          <SelectedTaskContext.Provider
            value={{ selectedTask: [], setSelectedTask }}
          >
            <ToDoList />
          </SelectedTaskContext.Provider>
        </IDContext.Provider>
      </QueryClientProvider>
    );
    expect(useQuery).toHaveBeenCalled();
    expect(screen.getByText("No Assigned Tasks")).toBeOnTheScreen();
  });
  it("Renders and fetches tasks", async () => {
    const data = [
      {
        Unit: "G",
        Type: "ELE",
        Description: "Read Needed",
        completed: false,
      },
    ];
    global.fetch = jest.fn().mockResolvedValueOnce({
      json: jest.fn().mockResolvedValueOnce(data),
    });

    const queryClient = new QueryClient();
    const mockData = [
      {
        Unit: "G",
        Type: "ELE",
        Description: "Read Needed",
        completed: false,
      },
    ];
    await act(() => {
      useQuery.mockReturnValue({
        data: mockData,
        isLoading: false,
        isError: false,
      });
    });

    const setSelectedTask = jest.fn();

    const showToggleButton = jest.fn();
    render(
      <QueryClientProvider client={queryClient}>
        <IDContext.Provider value={{ ID: "1" }}>
          <SelectedTaskContext.Provider
            value={{ selectedTask: [], setSelectedTask }}
          >
            <ToDoList showToggleButton={showToggleButton} />
          </SelectedTaskContext.Provider>
        </IDContext.Provider>
      </QueryClientProvider>
    );
    expect(useQuery).toHaveBeenCalled();
    expect(screen).toMatchSnapshot();
    await act(() => {
      const selectableTask = screen.getByTestId("selectTask");
      expect(selectableTask).toBeOnTheScreen();
      fireEvent.press(selectableTask);
    });
  });
  it("Renders Fetching Error", async () => {
    const queryClient = new QueryClient();
    const mockData = [
      {
        Unit: "G",
        Type: "ELE",
        Description: "Read Needed",
        completed: false,
      },
    ];
    await act(() => {
      useQuery.mockReturnValueOnce({
        data: mockData,
        isLoading: false,
        isError: true,
      });
    });

    const setSelectedTask = jest.fn();
    render(
      <QueryClientProvider client={queryClient}>
        <IDContext.Provider value={{ ID: "1" }}>
          <SelectedTaskContext.Provider
            value={{ selectedTask: [], setSelectedTask }}
          >
            <ToDoList />
          </SelectedTaskContext.Provider>
        </IDContext.Provider>
      </QueryClientProvider>
    );
    expect(useQuery).toHaveBeenCalled();
    expect(screen.getByText("Error fetching data")).toBeOnTheScreen();
  });

  it("Renders Activity Indicator", async () => {
    const data = [
      {
        Unit: "G",
        Type: "ELE",
        Description: "Read Needed",
        completed: false,
      },
    ];
    global.fetch = jest.fn().mockResolvedValueOnce({
      json: jest.fn().mockResolvedValueOnce(data),
    });

    const queryClient = new QueryClient();

    await act(() => {
      useQuery.mockReturnValue({
        data: undefined,
        isLoading: true,
        isError: false,
      });
    });

    const setSelectedTask = jest.fn();

    const showToggleButton = jest.fn();
    render(
      <QueryClientProvider client={queryClient}>
        <IDContext.Provider value={{ ID: "1" }}>
          <SelectedTaskContext.Provider
            value={{ selectedTask: [], setSelectedTask }}
          >
            <ToDoList showToggleButton={showToggleButton} />
          </SelectedTaskContext.Provider>
        </IDContext.Provider>
      </QueryClientProvider>
    );
    expect(useQuery).toHaveBeenCalled();
  });
});
