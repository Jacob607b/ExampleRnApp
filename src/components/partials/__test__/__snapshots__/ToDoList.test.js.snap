// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ToDoList Tests Renders and fetches tasks 1`] = `
<View
  animatedStyle={
    Object {
      "value": Object {},
    }
  }
  collapsable={false}
  exiting={[Function]}
  style={
    Object {
      "color": "#4D626C",
    }
  }
  testID="toDoList"
>
  <Modal
    hardwareAccelerated={false}
    onRequestClose={[Function]}
    presentationStyle="overFullScreen"
    transparent={true}
    visible={true}
  >
    <View
      style={
        Object {
          "alignItems": "center",
          "backgroundColor": "rgba(0, 0, 0, 0.5)",
          "flex": 1,
          "justifyContent": "center",
        }
      }
    >
      <View
        style={
          Object {
            "alignItems": "center",
            "backgroundColor": "#f2f2f2",
            "flexDirection": "column",
            "height": "60%",
            "justifyContent": "center",
            "padding": 20,
            "textAlign": "center",
            "width": "100%",
          }
        }
      >
        <View
          style={
            Object {
              "alignItems": "center",
              "backgroundColor": "#008dc9",
              "height": "10%",
              "justifyContent": "center",
              "marginBottom": 5,
              "marginTop": 5,
              "width": "100%",
            }
          }
        >
          <Text
            style={
              Object {
                "color": "white",
                "fontSize": 15,
                "fontWeight": "bold",
              }
            }
          >
            Please Select A Task
          </Text>
        </View>
        <RCTScrollView
          CellRendererComponent={[Function]}
          animatedStyle={
            Object {
              "value": Object {},
            }
          }
          collapsable={false}
          data={
            Array [
              Object {
                "Description": "Read Needed",
                "Type": "ELE",
                "Unit": "G",
                "completed": false,
              },
            ]
          }
          entering={
            FadeInDown {
              "build": [Function],
              "callbackV": undefined,
              "dampingRatioV": undefined,
              "dampingV": undefined,
              "delayV": undefined,
              "durationV": undefined,
              "easingV": undefined,
              "initialValues": undefined,
              "massV": 1,
              "overshootClampingV": undefined,
              "randomizeDelay": false,
              "reduceMotionV": "system",
              "restDisplacementThresholdV": undefined,
              "restSpeedThresholdV": undefined,
              "rotateV": undefined,
              "stiffnessV": undefined,
              "type": [Function],
            }
          }
          getItem={[Function]}
          getItemCount={[Function]}
          keyExtractor={[Function]}
          onContentSizeChange={[Function]}
          onLayout={[Function]}
          onMomentumScrollBegin={[Function]}
          onMomentumScrollEnd={[Function]}
          onScroll={[Function]}
          onScrollBeginDrag={[Function]}
          onScrollEndDrag={[Function]}
          removeClippedSubviews={false}
          renderItem={[Function]}
          scrollEventThrottle={1}
          stickyHeaderIndices={Array []}
          style={
            Object {
              "flex": 1,
            }
          }
          viewabilityConfigCallbackPairs={Array []}
        >
          <View>
            <View
              animatedStyle={
                Object {
                  "value": Object {},
                }
              }
              collapsable={false}
              onLayout={[Function]}
              style={Object {}}
            >
              <RCTScrollView>
                <View>
                  <View
                    accessibilityState={
                      Object {
                        "busy": undefined,
                        "checked": undefined,
                        "disabled": undefined,
                        "expanded": undefined,
                        "selected": undefined,
                      }
                    }
                    accessibilityValue={
                      Object {
                        "max": undefined,
                        "min": undefined,
                        "now": undefined,
                        "text": undefined,
                      }
                    }
                    accessible={true}
                    collapsable={false}
                    focusable={true}
                    onClick={[Function]}
                    onResponderGrant={[Function]}
                    onResponderMove={[Function]}
                    onResponderRelease={[Function]}
                    onResponderTerminate={[Function]}
                    onResponderTerminationRequest={[Function]}
                    onStartShouldSetResponder={[Function]}
                    style={
                      Object {
                        "opacity": 1,
                      }
                    }
                    testID="selectTask"
                  >
                    <View
                      style={
                        Object {
                          "alignItems": "center",
                          "backgroundColor": "#3ba1c5",
                          "borderColor": "grey",
                          "borderRadius": 20,
                          "borderWidth": 1,
                          "justifyContent": "center",
                          "marginBottom": 10,
                          "marginRight": 200,
                          "paddingBottom": 50,
                          "width": "100%",
                        }
                      }
                    >
                      <Text
                        style={
                          Object {
                            "color": "#fbfdf6",
                            "fontSize": 20,
                            "fontWeight": "bold",
                            "marginTop": 10,
                          }
                        }
                      >
                        Unit: 
                        G
                      </Text>
                      <Text
                        style={
                          Object {
                            "color": "#fbfdf6",
                            "fontSize": 20,
                            "fontWeight": "bold",
                            "marginTop": 10,
                          }
                        }
                      >
                        Task: 
                        ELE
                         
                        Read Needed
                      </Text>
                    </View>
                  </View>
                </View>
              </RCTScrollView>
            </View>
          </View>
        </RCTScrollView>
        <View
          accessibilityState={
            Object {
              "busy": undefined,
              "checked": undefined,
              "disabled": undefined,
              "expanded": undefined,
              "selected": undefined,
            }
          }
          accessibilityValue={
            Object {
              "max": undefined,
              "min": undefined,
              "now": undefined,
              "text": undefined,
            }
          }
          accessible={true}
          collapsable={false}
          focusable={true}
          onClick={[Function]}
          onResponderGrant={[Function]}
          onResponderMove={[Function]}
          onResponderRelease={[Function]}
          onResponderTerminate={[Function]}
          onResponderTerminationRequest={[Function]}
          onStartShouldSetResponder={[Function]}
          style={
            Object {
              "marginBottom": 5,
              "marginTop": 10,
              "opacity": 1,
            }
          }
          testID="Dismiss"
        >
          <View
            collapsable={false}
            style={
              Object {
                "backgroundColor": "rgb(247, 243, 249)",
                "borderRadius": 20,
                "shadowColor": "#000",
                "shadowOffset": Object {
                  "height": 1,
                  "width": 0,
                },
                "shadowOpacity": 0.15,
                "shadowRadius": 3,
              }
            }
            testID="button-container-outer-layer"
          >
            <View
              collapsable={false}
              style={
                Object {
                  "backgroundColor": "rgb(247, 243, 249)",
                  "borderColor": "transparent",
                  "borderRadius": 20,
                  "borderStyle": "solid",
                  "borderWidth": 0,
                  "flex": undefined,
                  "minWidth": "auto",
                  "shadowColor": "#000",
                  "shadowOffset": Object {
                    "height": 1,
                    "width": 0,
                  },
                  "shadowOpacity": 0.3,
                  "shadowRadius": 1,
                }
              }
              testID="button-container"
            >
              <View
                accessibilityRole="button"
                accessibilityState={
                  Object {
                    "busy": undefined,
                    "checked": undefined,
                    "disabled": true,
                    "expanded": undefined,
                    "selected": undefined,
                  }
                }
                accessibilityValue={
                  Object {
                    "max": undefined,
                    "min": undefined,
                    "now": undefined,
                    "text": undefined,
                  }
                }
                accessible={true}
                collapsable={false}
                focusable={true}
                onBlur={[Function]}
                onClick={[Function]}
                onFocus={[Function]}
                onResponderGrant={[Function]}
                onResponderMove={[Function]}
                onResponderRelease={[Function]}
                onResponderTerminate={[Function]}
                onResponderTerminationRequest={[Function]}
                onStartShouldSetResponder={[Function]}
                style={
                  Array [
                    Object {
                      "overflow": "hidden",
                    },
                    Object {
                      "borderRadius": 20,
                    },
                  ]
                }
                testID="button"
              >
                <View
                  style={
                    Array [
                      Object {
                        "alignItems": "center",
                        "flexDirection": "row",
                        "justifyContent": "center",
                      },
                      undefined,
                    ]
                  }
                >
                  <Text
                    numberOfLines={1}
                    selectable={false}
                    style={
                      Array [
                        Object {
                          "textAlign": "left",
                        },
                        Object {
                          "color": "rgba(28, 27, 31, 1)",
                          "writingDirection": "ltr",
                        },
                        Array [
                          Object {
                            "fontFamily": "System",
                            "fontSize": 14,
                            "fontWeight": "500",
                            "letterSpacing": 0.1,
                            "lineHeight": 20,
                          },
                          Array [
                            Object {
                              "marginHorizontal": 16,
                              "marginVertical": 9,
                              "textAlign": "center",
                            },
                            false,
                            Object {
                              "marginHorizontal": 24,
                              "marginVertical": 10,
                            },
                            Object {
                              "marginHorizontal": 8,
                            },
                            false,
                            Object {
                              "color": "rgba(103, 80, 164, 1)",
                              "fontFamily": "System",
                              "fontSize": 14,
                              "fontWeight": "500",
                              "letterSpacing": 0.1,
                              "lineHeight": 20,
                            },
                            undefined,
                          ],
                        ],
                      ]
                    }
                    testID="button-text"
                  >
                    Close
                  </Text>
                </View>
              </View>
            </View>
          </View>
        </View>
      </View>
    </View>
  </Modal>
</View>
`;
